driver/src/price_estimation/dexag/mod.rs

```diff
@@ -1 +1,178 @@
 mod api;
+
+use super::{PriceSource, Token};
+use crate::models::TokenId;
+use anyhow::{anyhow, Result};
+use api::{DexagApi, DexagHttpApi};
+use std::collections::HashMap;
+
+pub struct DexagClient<Api> {
+    api: Api,
+    // Maps Token::symbol to Token.
+    // This is cached in the struct because we don't expect it to change often.
+    tokens: HashMap<String, api::Token>,
+    stable_coin: api::Token,
+}
+
+impl DexagClient<DexagHttpApi> {
+    /// Create a DexagClient using DexagHttpApi as the api implementation.
+    #[allow(dead_code)]
+    pub fn new() -> Result<Self> {
+        let api = DexagHttpApi::new()?;
+        Self::with_api(api)
+    }
+}
+
+impl<Api> DexagClient<Api>
+where
+    Api: DexagApi,
+{
+    pub fn with_api(api: Api) -> Result<Self> {
+        // We need to return prices in OWL but Dexag does not track it. OWL tracks
+        // USD so we use another stable coin as an approximate USD price.
+        const STABLE_COIN: &str = "DAI";
+
+        let tokens = api.get_token_list()?;
+        let mut tokens: HashMap<String, api::Token> = tokens
+            .into_iter()
+            .map(|token| (token.symbol.clone(), token))
+            .collect();
+        let stable_coin = tokens.remove(STABLE_COIN).ok_or_else(|| {
+            anyhow!(
+                "dexag exchange does not track our stable coin {}",
+                STABLE_COIN
+            )
+        })?;
+
+        Ok(Self {
+            api,
+            tokens,
+            stable_coin,
+        })
+    }
+}
+
+impl<Api> PriceSource for DexagClient<Api>
+where
+    Api: DexagApi,
+{
+    fn get_prices(&self, tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
+        Ok(tokens
+            .iter()
+            .filter_map(|token| {
+                let stable_coin_price = if token.symbol() == self.stable_coin.symbol {
```

Comment:
I'm a bit confused by the name. Is this name supposed to imply that this price is always in reference to the stable coin? Should it be called `priceInOwl` or something like that

Reply:
> Is this name supposed to imply that this price is always in reference to the stable coin?

That is the case. The reason for why there is a stable coin is explained in a comment earlier in the file.